src/datasets/arrow_dataset.py

```diff
@@ -3778,10 +3844,25 @@ def concatenate_datasets(
 
             .. versionadded:: 1.6.0
     """
-    if axis == 0 and not all([dset.features.type == dsets[0].features.type for dset in dsets]):
-        raise ValueError("Features must match for all datasets")
-    elif axis == 1 and not all([dset.num_rows == dsets[0].num_rows for dset in dsets]):
-        raise ValueError("Number of rows must match for all datasets")
+    # Ignore datasets with no rows
+    if any(dset.num_rows > 0 for dset in dsets):
+        dsets = [dset for dset in dsets if dset.num_rows > 0]
+    else:
+        # Return first dataset if all datasets are empty
+        return dsets[0]
+
+    # Perform checks (and a potentional cast if axis=0)
+    if axis == 0:
+        _check_if_features_can_be_aligned([dset.features for dset in dsets])
+        aligned_features = _align_features([dset.features for dset in dsets])
+        dsets = [
+            dset.cast(features) if dset.features != features else dset
```

Comment:
`dset.cast` is costly since it calls `map`. Is the issue just about reordering the features ?
If so maybe we could have an additional (private) method that does that instead of calling `map` ?